### PR TITLE
feat(components): main components PE-1283

### PR DESCRIPTION
**Pull request by karlprieb**
This commit implements the following components: box, countdown, footer, header, rank and sidebar
An index page mounting all the components

Implements jest with testing-library

image
image
image

### CODE REVIEW

File: components/rank/index.tsx [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+	if (cent >= 10 && cent <= 20) return 'th';
+	const dec = indicator % 10;
+	if (dec === 1) return 'st';
+	if (dec === 2) return 'nd';
+	if (dec === 3) return 'rd';
+	return 'th';
+};
+
+const calculateByteSize = (bytes: number): string => {
+	const size = byteSize(bytes, { units: 'iec' });
```
Comment (fedellen) [inline components/rank/index.tsx:98]: What exactly is this package doing here? Are we converting from decimal/binary or something? Arweave and the ArDrive apps only handle bytes in binary

We do have a `formatBytes` function in Core that performs something similar. We could just copy paste a version of that function into a utils file rather than having a whole dependency added here.


```ts
// Format byte size to something nicer.  This is minified...
export function formatBytes(bytes: number): string {
	const marker = 1024; // Change to 1000 if required
	const decimal = 3; // Change as required
	const kiloBytes = marker; // One Kilobyte is 1024 bytes
	const megaBytes = marker * marker; // One MB is 1024 KB
	const gigaBytes = marker * marker * marker; // One GB is 1024 MB
	// const teraBytes = marker * marker * marker * marker; // One TB is 1024 GB

	// return bytes if less than a KB
	if (bytes < kiloBytes) return `${bytes} Bytes`;
	// return KB if less than a MB
	if (bytes < megaBytes) return `${(bytes / kiloBytes).toFixed(decimal)} KB`;
	// return MB if less than a GB
	if (bytes < gigaBytes) return `${(bytes / megaBytes).toFixed(decimal)} MB`;
	// return GB if less than a TB
	return `${(bytes / gigaBytes).toFixed(decimal)} GB`;
}
```

File: components/rank/index.tsx [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+	}
+	if (change < 0) {
+		Arrow = ArrowDown;
+	}
+
+	return (
+		<>
+			{Math.abs(change)}% {Arrow ? <Arrow /> : null}
+		</>
+	);
```
Comment (fedellen) [inline components/rank/index.tsx:115]: Consider handling the variable conditions in one scope:

```ts
	const Arrow = (() => {
		if (change > 0) {
			return ArrowUp;
		}
		if (change < 0) {
			return ArrowDown;
		}

		return null;
	})();

	return (
		<>
			{Math.abs(change)}% {Arrow}
		</>
	);

File: components/rank/index.tsx [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+		}
+	}
+`;
+
+const formatWalletAddress = (wallet: string): string => {
+	const first5Chars = wallet.slice(0, 5);
+	const last5Chars = wallet.slice(-5);
+
+	return `${first5Chars}...${last5Chars}`;
+};
```
Comment (fedellen) [inline components/rank/index.tsx:84]: I really like the way these variable names describe the string manipulation being used here, very helpful.

I do think all of these functions should be moved into the utils folder somewhere though. No need to pollute the components with all of this logic.

File: components/rank/index.tsx [hunk 4]
```diff
@@ -0,0 +1,171 @@
+import styled from 'styled-components';
+import byteSize from 'byte-size';
+
+import { ArrowDown, ArrowUp } from '@components/icons';
+
+import device from '@utils/media_queries';
+import type { Data } from '@utils/dataType';
+
+const RankWrapper = styled.div`
```
Comment (fedellen) [inline components/rank/index.tsx:9]: Consider splitting large css blocks into a `*.style.tsx` file

File: components/rank/index.tsx [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+	const first5Chars = wallet.slice(0, 5);
+	const last5Chars = wallet.slice(-5);
+
+	return `${first5Chars}...${last5Chars}`;
+};
+
+const positionIndicator = (position: number): string => {
+	const indicator = Math.abs(position);
+	const cent = indicator % 100;
+	if (cent >= 10 && cent <= 20) return 'th';
```
Comment (fedellen) [inline components/rank/index.tsx:89]: style nit: We usually prefer adding the brackets to these `if` cases that simply return so its easier for the code reader to follow the returns through indentation

I believe @arielmelendez called me out on this one before 😅

File: components/countdown/index.tsx [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+	const nextSunday = new Date(today);
+
+	do {
+		nextSunday.setDate(nextSunday.getDate() + 1);
+	} while (nextSunday.getDay() !== sundayNumber);
+
+	return new Date(nextSunday.toLocaleString('en-US', { timeZone: timezone }));
+};
+
+const getValues = (countDown: number): string[] => {
```
Comment (matibat) [inline components/countdown/index.tsx:24]: It could be convenient to have a fixed size array as return type
```suggestion
const getValues = (countDown: number): [string, string, string, string] => {
```

File: components/footer/index.tsx [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+		}
+	}
+`;
+
+const Footer = ({ lastUpdated }: { lastUpdated: number }): JSX.Element => {
+	const fromLastUpdated = dayjs(new Date()).to(dayjs(lastUpdated));
+
+	return (
+		<FooterWrapper>
+			<a href='#'>Rules</a>
```
Comment (matibat) [inline components/footer/index.tsx:75]: TODO: add links for the rules and to share on TW

File: mocks/output.json [hunk 8]
```diff
@@ -0,0 +1,226 @@
...
+				"fileCount": 10,
+				"rankPosition": 49,
+				"tokensEarned": 100
+			},
+			"total": {
+				"byteCount": 100,
+				"changeInPercentage": {
+					"24h": -50,
+					"7d": 40
+				},
```
Comment (matibat) [inline mocks/output.json:34]: I'm currently not tracking the change in percentage of the total. But I think we don't need it :thinking:

File: mocks/output.json [hunk 9]
```diff
@@ -0,0 +1,226 @@
...
+				"ka_hAAWru0qzK0KprCbvHX5J6QJtkxEiDOHMM1kNB1f": 100,
+				"ka_hAAWru0qzK0KprCbvHX5J6QJtkxEiDOHMM1kNB1g": 100,
+				"ka_hAAWru0qzK0KprCbvHX5J6QJtkxEiDOHMM1kNB1h": 100,
+				"ka_hAAWru0qzK0KprCbvHX5J6QJtkxEiDOHMM1kNB1j": 100
+			},
+			"streakRewards": {
+				"ka_hAAWru0qzK0KprCbvHX5J6QJtkxEiDOHMM1kNB1j": 100,
+				"ka_hAAWru0qzK0KprCbvHX5J6QJtkxEiDOHMM1kNB1h": 100
+			}
+		}
```
Comment (matibat) [inline mocks/output.json:224]: Missing `ranks.total`

File: pages/_app.tsx [hunk 10]
```diff
@@ -1,10 +1,12 @@
...
+import './global_style.css';
 import type { AppProps, NextWebVitalsMetric } from 'next/app';
 
 function MyApp({ Component, pageProps }: AppProps) {
 	return <Component {...pageProps} />;
 }
 
-export function reportWebVitals(metric: NextWebVitalsMetric) {
+export function reportWebVitals(metric: NextWebVitalsMetric): void {
 	if (process.env.NODE_ENV === 'development') console.log(metric);
```
Comment (matibat) [inline pages/_app.tsx:10]: Nit: single-line conditional

**Review by matibat — COMMENTED** [review]
We'll need a `favicon.ico`

File: components/footer/index.tsx (hunk 7 above)
Comment (matibat) [reply components/footer/index.tsx:75]: Rules handled at PE-1288

File: components/rank/index.tsx (hunk 3 above)
Comment (karlprieb) [reply components/rank/index.tsx]: Fixed in https://github.com/ardriveapp/inferno-leaderboard/commit/1f3a4d84239961d6c2d893dd393b8b23424d3b7b

File: components/rank/index.tsx (hunk 4 above)
Comment (karlprieb) [reply components/rank/index.tsx:8]: We have less than 70 lines of styles here. I can't see why it could be a problem.

**Comment by karlprieb** [comment]
> We'll need a `favicon.ico`
We have it already 
https://github.com/ardriveapp/inferno-leaderboard/blob/1f3a4d84239961d6c2d893dd393b8b23424d3b7b/public/favicon.ico

File: components/rank/index.tsx (hunk 2 above)
Comment (karlprieb) [reply components/rank/index.tsx]: Fixed in 4cd85a595348399b9c5bbd7500a281a9c9c600dc

File: components/rank/index.tsx (hunk 5 above)
Comment (karlprieb) [reply components/rank/index.tsx]: Fixed in 697b5596c0cd6d7fa415ddcb06f8d1c913b899a1

File: components/main/index.tsx [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+	height: 100%;
+	overflow-y: scroll;
+
+	@media ${device.tablet} {
+		padding: 0 1rem 0 2rem;
+		margin-right: 1rem;
+
+		& {
+			scrollbar-color: #2c2c2c transparent;
+		}
```
[the remark this answers is not in the document]
Comment (karlprieb) [reply components/main/index.tsx]: fixed in https://github.com/ardriveapp/inferno-leaderboard/pull/2/commits/d0c3366614aa59f964c69062c22da6dea86309fe

File: components/countdown/index.tsx (hunk 6 above)
Comment (karlprieb) [reply components/countdown/index.tsx]: fixed in https://github.com/ardriveapp/inferno-leaderboard/pull/2/commits/cf627958c0eec81cb5cb6e6a9817b75b188c5975

File: components/box/index.tsx [hunk 12]
```diff
@@ -0,0 +1,72 @@
+import styled, { css } from 'styled-components';
+
+import { ArrowUp, ArrowDown } from '@components/icons';
+
+import device from '@utils/media_queries';
+
+export const Wrapper = styled.div`
```
[the remark this answers is not in the document]
Comment (karlprieb) [reply components/box/index.tsx]: https://github.com/ardriveapp/inferno-leaderboard/pull/2/commits/8895788c1d67d4a3a6b98d802c8a0fefba61564c

File: mocks/output.json (hunk 8 above)
Comment (karlprieb) [reply mocks/output.json]: https://github.com/ardriveapp/inferno-leaderboard/pull/2/commits/033d2bf4d4e81b03a8f56ec6fdda9d9f220205eb

File: mocks/output.json (hunk 9 above)
Comment (karlprieb) [reply mocks/output.json:275]: https://github.com/ardriveapp/inferno-leaderboard/pull/2/commits/f378978869d585f547ae4ca9fbd3cc220f697804

File: components/countdown/index.tsx [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+	today.setMinutes(0);
+	today.setHours(0);
+
+	const nextSunday = new Date(today);
+
+	do {
+		nextSunday.setDate(nextSunday.getDate() + 1);
+	} while (nextSunday.getDay() !== sundayNumber);
+
+	return new Date(nextSunday.toLocaleString('en-US', { timeZone: timezone }));
```
Comment (fedellen) [inline components/countdown/index.tsx:21]: This function is a little hard to follow. It could use comments or more descriptive variable names. 

And there could be some improvement on the locality of the variables, for eg:

```suggestion
	const now = new Date();
	const today = new Date(now);
	
	today.setMilliseconds(0);
	today.setSeconds(0);
	today.setMinutes(0);
	today.setHours(0);

	const nextSunday = new Date(today);

	const sundayNumber = 0;
	do {
		nextSunday.setDate(nextSunday.getDate() + 1);
	} while (nextSunday.getDay() !== sundayNumber);


	const timezone = 'America/New_York';
	return new Date(nextSunday.toLocaleString('en-US', { timeZone: timezone }));
```

File: components/countdown/index.tsx [hunk 14]
```diff
@@ -0,0 +1,65 @@
...
+	});
+
+	const minutes = Math.floor((countDown % (1000 * 60 * 60)) / (1000 * 60)).toLocaleString('en-US', {
+		minimumIntegerDigits: 2,
+		useGrouping: false,
+	});
+
+	const seconds = Math.floor((countDown % (1000 * 60)) / 1000).toLocaleString('en-US', {
+		minimumIntegerDigits: 2,
+		useGrouping: false,
```
Comment (fedellen) [inline components/countdown/index.tsx:39]: These static numbers should be extracted into their own variables and perhaps hoisted out of this function if they're being used. e.g:

```ts
const oneSecond = 1000
const oneMinute = oneSecond * 60
const oneHour = oneMinute * 60
const oneDay = oneHour * 24
```

File: components/countdown/index.tsx [hunk 15]
```diff
@@ -0,0 +1,65 @@
...
+};
+
+const Countdown = (): JSX.Element => {
+	const targetDate = getTargetDate().getTime();
+
+	const [countDown, setCountDown] = useState(targetDate - new Date().getTime());
+
+	useEffect(() => {
+		const interval = setInterval(() => {
+			setCountDown(targetDate - new Date().getTime());
```
Comment (fedellen) [inline components/countdown/index.tsx:52]: Could improve readability for those unfamiliar with JavaScript `Date` with a variable here:

```suggestion
        const currentTime = new Date().getTime()
	const [countDown, setCountDown] = useState(targetDate - currentTime);

	useEffect(() => {
		const interval = setInterval(() => {
			setCountDown(targetDate - currentTime);
```

File: components/footer/index.test.tsx [hunk 16]
```diff
@@ -0,0 +1,15 @@
+import { render, screen } from '@testing-library/react';
+import Footer from '@components/footer';
+
+describe('Box', () => {
```
Comment (fedellen) [inline components/footer/index.test.tsx:4]:
```suggestion
describe('Footer', () => {
```

File: components/footer/index.test.tsx [hunk 17]
```diff
@@ -0,0 +1,15 @@
...
+	it('renders last updated', () => {
+		const minutesAgo = new Date().getTime() - 1800000;
```
Comment (fedellen) [inline components/footer/index.test.tsx:6]:
```suggestion
		const thirtyMinutesAgo = new Date().getTime() - 1_800_000;
```

File: components/footer/index.tsx (hunk 7 above)
Comment (fedellen) [reply components/footer/index.tsx:75]: Probably not for this ticket, but a reminder to ask Lucas about a mouse hover color for this link. And any other clickable components

File: components/footer/index.tsx [hunk 18]
```diff
@@ -0,0 +1,87 @@
...
+		}
+
+		& > svg {
+			width: 27px;
+		}
+	}
+`;
+
+const Footer = ({ lastUpdated }: { lastUpdated: number }): JSX.Element => {
+	const fromLastUpdated = dayjs(new Date()).to(dayjs(lastUpdated));
```
Comment (fedellen) [inline components/footer/index.tsx:71]: Might want to leave a comment on what `dayjs` is doing here. Seems to be quite a few things happening in 1 line of code

File: components/main/index.tsx [hunk 19]
```diff
@@ -0,0 +1,70 @@
...
+	grid-template-rows: 100px 1fr;
+	gap: 1rem;
+
+	@media ${device.tablet} {
+		grid-template-columns: 70% 1fr;
+		grid-template-rows: auto;
+	}
+`;
+
+const RankWrapper = styled.div`
```
Comment (fedellen) [inline components/main/index.tsx:21]: html nit: You could consider `section` element instead of `div` for `RankWrapper` and `SidebarWrapper`

I'm impressed with the overall the HTML though (:

File: components/sidebar/index.tsx [hunk 20]
```diff
@@ -0,0 +1,59 @@
...
+const Sidebar = ({
+	uploaders,
+	data,
+	files,
+	streakers,
+}: {
+	uploaders?: number;
+	data?: number;
+	files?: number;
+	streakers?: number;
```
Comment (fedellen) [inline components/sidebar/index.tsx:44]: What is the use case that this component could be rendered without any of these props? Or rather, why are they all optional here?

**Review by fedellen — APPROVED** [review]
Left some readability nits and suggestions.

The HTML / CSS side is looking excellent and everything is functioning well. Awesome work!

File: components/sidebar/index.tsx (hunk 20 above)
Comment (karlprieb) [reply components/sidebar/index.tsx:44]: Ouch... I really don't know ahhaha
I started this component thinking on getting all the data transformed before and just passed to it, so I could show and hide boxes for mobile/desktop, but on #2 I changed my mind and started to address all inside the componenet.

File: components/countdown/index.tsx (hunk 13 above)
Comment (karlprieb) [reply components/countdown/index.tsx]: Great readability suggestion!
Applied at https://github.com/ardriveapp/inferno-leaderboard/pull/2/commits/805e1f9541a9050fc53f84aa5c6f2d61d083ef85

File: components/countdown/index.tsx (hunk 14 above)
Comment (karlprieb) [reply components/countdown/index.tsx]: 👍 https://github.com/ardriveapp/inferno-leaderboard/pull/2/commits/805e1f9541a9050fc53f84aa5c6f2d61d083ef85

File: components/countdown/index.tsx (hunk 15 above)
Comment (karlprieb) [reply components/countdown/index.tsx]: I made a function to get the current time as it changes on every second
https://github.com/ardriveapp/inferno-leaderboard/pull/2/commits/805e1f9541a9050fc53f84aa5c6f2d61d083ef85

File: components/footer/index.tsx (hunk 18 above)
Comment (karlprieb) [reply components/footer/index.tsx]: https://github.com/ardriveapp/inferno-leaderboard/pull/2/commits/aba04df74eaba97abe07cb992c659785ddbfc5c2

File: components/main/index.tsx (hunk 19 above)
Comment (karlprieb) [reply components/main/index.tsx]: https://github.com/ardriveapp/inferno-leaderboard/pull/2/commits/8bf56c0a9a8e4a51c59860a472163fddc32061c9

**Review by fedellen — APPROVED** [review]
